README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# semiBRM: Semiparametric Binary Response Models in R

This **R** package **semiBRM** offers an implementation of single-index
semiparametric binary response models, theorized in a seminal work in
the semiparametric econometrics literature [Klein and Spady
(1993)](https://doi.org/10.2307/2951556).

This is built upon [Rcpp](http://www.rcpp.org) along with OpenMP, which
parallelizes computation of nonparametric conditional expectations over
given data points. This improves computation efficiency enormously in
parameter estimation, which is one of the main features of this package.

The package is still in development, with the goal of making it as easy
to employ [Klein and Spady (1993)](https://doi.org/10.2307/2951556) as
to run Probit or Logistic regressions in **R**.

## Identifiable Set of Parameters

In a single-index semiparametric approach, not all parameters are
identifiable. First of all, intercept cannot be estimated. Secondly, the
coefficients can be estimated as ratios to a “basis” coefficient. This
package sets the coefficient of the first explanatory variable as the
basis so that its coefficient is normalized to 1, and those of the rest
of the variables are estimated conformably to it.

In the context of binary response models, or nonlinear models more
broadly, this is not problematic at all, as the coefficients themselves
have little interpretation. As the matter of fact, the identifiable set
of parameters in this approach can correctly estimate the conditional
probability of Pr(Y=1|X), which would be the quantity of interest in
many cases.

## Installation

The development version can be installed from GitHub:

``` r
library(devtools)
install_github(repo="hk599/semiBRM")
```

### This version was built with:

  - R 4.0.4
  - rtools40
  - maxLik 1.4-6
  - Rcpp 1.0.6
  - RcppArmadillo 0.10.2.2.0

## Example

### Setup

``` r
library(semiBRM)
set.seed(20190815) # for reproduction of results
```

### Data Generating Process

``` r
## data generating process
N <- 1500L
X1 <- rnorm(N)
X2 <- (X1 + 2*rnorm(N))/sqrt(5) + 1
X3 <- rnorm(N)^2/sqrt(2)
X <- cbind(X1, X2, X3)
beta <- c(2, 2, -1, -1) # this is the original set of coefficients
V <- as.vector(cbind(X, 1)%*%beta)
Y <- ifelse(V >= rnorm(N), 1L, 0L)
```

### Parameter Estimation

``` r
## estimands: the rescaled coefficients by the first coefficient excluding intercept
coefs_true <- c(1, -.5)
data <- data.frame(Y, X1, X2, X3)

## Klein and Spady (1993): semiparametric approach
semi <- semiBRM(Y ~ X1 + X2 + X3, data = data, control = list(iterlim=50))
coefs_semi <- coef(semi)

## Probit: parametric approach
probit <- glm(Y ~ X1 + X2 + X3, family = binomial(link = "probit"), data = data)
coefs_probit <- probit$coefficients[-1L][-1L]/probit$coefficients[-1L][1L]

## formatted print
{
    cat(sprintf("        %7s %7s %7s\n", "True", "Probit", "Semi"))
    cat(sprintf("parm 1: %7.4f %7.4f %7.4f\n", coefs_true[1L], coefs_probit[1L], coefs_semi[1L]))
    cat(sprintf("parm 2: %7.4f %7.4f %7.4f\n", coefs_true[2L], coefs_probit[2L], coefs_semi[2L]))
}
#>            True  Probit    Semi
#> parm 1:  1.0000  0.9296  0.9430
#> parm 2: -0.5000 -0.4069 -0.4103
```

### In-Sample Prediction

``` r
## in-sample conditional probability
in_prob_true <- pnorm(V)
in_prob_semi <- predict(semi)
in_prob_probit <- fitted(probit)

## formatted print
target <- sample.int(N, size = 10L)
{
    cat(sprintf("%7s %8s %8s %8s %12s\n", "Obs.", "True", "Probit", "Semi", "non.endpoint") )
    for (i in target){
        cat(sprintf("[%04d]: %8.6f %8.6f %8.6f %12s\n",
                    i, in_prob_true[i], in_prob_probit[i], in_prob_semi$prob[i], in_prob_semi$non.endpoint[i]))
    }
}
#>    Obs.     True   Probit     Semi non.endpoint
#> [0034]: 0.999400 0.999262 0.991678         TRUE
#> [0514]: 0.898871 0.859331 0.774815         TRUE
#> [0480]: 0.995355 0.994697 0.970159         TRUE
#> [1285]: 0.004911 0.006525 0.034459         TRUE
#> [0065]: 0.990712 0.993818 0.966249         TRUE
#> [0966]: 0.896823 0.862113 0.780052         TRUE
#> [0558]: 0.997327 0.997560 0.981306         TRUE
#> [1001]: 0.000001 0.000000 0.000653         TRUE
#> [0843]: 0.998430 0.998133 0.984302         TRUE
#> [0381]: 0.781482 0.773341 0.697235         TRUE
```

### Out-of-Sample Prediction

``` r
## conditional probability at the means
Xbar <- colMeans(X)
newdata <- as.data.frame(as.list(Xbar))

## predictions
out_prob_true <- pnorm(as.vector(c(Xbar, 1)%*%beta))
out_prob_semi <- predict(semi, newdata, boot.se = TRUE)
out_prob_probit <- pnorm(as.vector(coef(probit)%*%c(1, Xbar)))

## standard errors of Probit
grad <- dnorm(as.vector(coef(probit)%*%c(1, Xbar))) * c(1, Xbar)
out_stde_probit <- sqrt(as.vector(crossprod(grad, vcov(probit))%*%grad))

## formatted print
{
    cat(sprintf("            %7s %7s %7s\n", "True", "Probit", "Semi"))
    cat(sprintf("Prob. Est.: %7.4f %7.4f %7.4f\n", out_prob_true, out_prob_probit, out_prob_semi$prob))
    cat(sprintf(" Std. Err.: %7s %7.4f %7.4f\n", "", out_stde_probit, out_prob_semi$boot.se))
}
#>                True  Probit    Semi
#> Prob. Est.:  0.5833  0.5792  0.5543
#>  Std. Err.:          0.0230  0.0221
```

### Average Marginal Effects

``` r
## marginal effects as difference between conditional probabilities with and without perturbation
delta <- sd(X1) # size of perturbation

me_true <- pnorm(as.vector(cbind(X1+delta, X2, X3, 1)%*%beta)) - pnorm(as.vector(cbind(X, 1)%*%beta))

## average marginal effects
av_me_true <- mean(me_true)
av_me_semi <- MarginalEffects(semi, variable = "X1", delta = delta)

## formatted print
{
    cat(sprintf("           %7s %7s \n", "True", "Semi"))
    cat(sprintf("Marg.Eff.: %7.4f %7.4f\n", av_me_true, av_me_semi[1L]))
    cat(sprintf("Std. Err.: %7s %7.4f\n", "", av_me_semi[2L]))
}
#>               True    Semi 
#> Marg.Eff.:  0.2021  0.2029
#> Std. Err.:          0.0050
```

### Quantile Marginal Effects

``` r
## percentile cutoffs
p.cutoffs <- c(1/4, 2/4, 3/4)

## group indicators
q_vals <- quantile(X1, p.cutoffs)
q1 <- ifelse(X1 <= q_vals[1L], 1L, 0L)
q2 <- ifelse(q_vals[1L] < X1 & X1 <= q_vals[2L], 1L, 0L)
q3 <- ifelse(q_vals[2L] < X1 & X1 <= q_vals[3L], 1L, 0L)
q4 <- ifelse(q_vals[3L] < X1, 1L, 0L)

## quantile marginal effects
q_me_true <- c("G1" = mean(me_true[q1==1]), "G2" = mean(me_true[q2==1]),
               "G3" = mean(me_true[q3==1]), "G4" = mean(me_true[q4==1]))

q_me_semi <- MarginalEffects(semi, variable = "X1", delta, p.cutoffs)

## formatted print
{
    cat("         ", sprintf(" %5s  ", paste0("G", 1:4)), fill = TRUE)
    cat("True    :", sprintf(" %.4f ", q_me_true), fill = TRUE)
    cat("Semi    :", sprintf(" %.4f ", q_me_semi[,1L]), fill = TRUE)
    cat("Std.Err.:", sprintf("(%.4f)", q_me_semi[,2L]), fill = TRUE)
}
#>               G1       G2       G3       G4  
#> True    :  0.2232   0.3175   0.2196   0.0481 
#> Semi    :  0.2186   0.3130   0.2248   0.0552 
#> Std.Err.: (0.0100) (0.0088) (0.0096) (0.0061)
```

## References

Eddelbuettel, D., & François, R. (2011). Rcpp: Seamless R and C++
integration. *Journal of Statistical Software*, *40(8)*, 1-18.
*<https://dirk.eddelbuettel.com/code/rcpp/Rcpp-introduction.pdf>*

Klein, R. W., & Spady, R. H. (1993). An Efficient Semiparametric
Estimator for Binary Response Models. *Econometrica*, *61(2)*, 387-421.
*<https://doi.org/10.2307/2951556>*.

Klein, R., & Vella, F. (2009). A Semiparametric Model for Binary
Response and Continuous Outcomes Under Index Heteroscedasticity.
*Journal of Applied Econometrics*, *24(5)*, 735-762.
*<https://doi.org/10.1002/jae.1064>*


NEWS.md
# semiBRM 0.0.1

## The 1st release to CRAN

* This offers a general framework to implement semiparametric binary response models in R.


tests/testthat.R
library(testthat)
library(semiBRM)

test_check("semiBRM")


tests/testthat/test-Rcpp-functions.R
# R functions
## univariate case
RKSCE <- function(Y, X, args, h)
{
    Nargs <- length(args)
    output <- rep(NaN, Nargs)

    for ( i in seq_len(Nargs) ){
        Z <- (X-args[i])/h
        Kerh <- stats::dnorm(Z)/h
        output[i] <- sum(Y*Kerh)/sum(Kerh)
    }

    output
}

## univariate case, leave-one-out
RKSCE_leave_one_out <- function(Y, X, h)
{
    Nargs <- length(X)
    output <- rep(NaN, Nargs)

    for ( i in seq_len(Nargs) ){
        Z <- (X[-i]-X[i])/h
        Kerh <- stats::dnorm(Z)/h
        output[i] <- sum(Y[-i]*Kerh)/sum(Kerh)
    }

    output
}

## multivariate case
RKMCE <- function(Y, X, args, h)
{
    Nargs <- nrow(args)
    output <- rep(NaN, Nargs)

    for ( i in seq_len(Nargs) ){
        D <- sweep(X, MARGIN = 2, STATS = args[i,], FUN = "-", check.margin = FALSE)
        Z <- sweep(D, MARGIN = 2, STATS = h, FUN = "/", check.margin = FALSE )
        Kerh <- sweep(stats::dnorm(Z), MARGIN = 2, STATS = h, FUN = "/", check.margin = FALSE)
        w <- apply(Kerh, 1, prod)
        output[i] <- sum(w*Y)/sum(w)
    }

    output
}

## multivariate case, leave-one-out
RKMCE_leave_one_out<- function(Y, X, h)
{
    Nargs <- nrow(X)
    output <- rep(NaN, Nargs)

    for ( i in seq_len(Nargs) ){
        D <- sweep(X[-i,], MARGIN = 2, STATS = X[i,], FUN = "-", check.margin = FALSE)
        Z <- sweep(D, MARGIN = 2, STATS = h, FUN = "/", check.margin = FALSE )
        Kerh <- sweep(stats::dnorm(Z), MARGIN = 2, STATS = h, FUN = "/", check.margin = FALSE)
        w <- apply(Kerh, 1, prod)
        output[i] <- sum(w*Y[-i])/sum(w)
    }

    output
}


# Test Rcpp functions
test_that("Rcpp functions should return the same results as those written in R.", {

    ## data generating process
    N <- 500L
    x1 <- stats::rnorm(N)
    x2 <- stats::rnorm(N)
    x3 <- stats::rnorm(N)
    x <- cbind(x1, x2, x3)
    v <- (x1 + x2 + x3)/sqrt(3)
    y <- v + rnorm(N)

    ## univariate case
    h_u <- stats::sd(v)*N^(-1/5)
    vargs <- stats::quantile(v)

    yhat_univ_1 <- RKSCE(y, v, vargs, h_u)
    yhat_univ_2 <- RKSCE_leave_one_out(y, v, h_u)

    ## multivariate case
    h_m <- apply(matrix(rnorm(150), 50, 3), 2, stats::sd)*N^(-1/7)
    xargs <- apply(matrix(rnorm(150), 50, 3), 2, stats::quantile)

    yhat_multi_1 <- RKMCE(y, x, xargs, h_m)
    yhat_multi_2 <- RKMCE_leave_one_out(y, x, h_m)

    ## testing
    expect_equal(GaussianNadarayaWatsonEstimator(v, y, h_u, vargs), yhat_univ_1)
    expect_equal(GaussianNadarayaWatsonEstimator(v, y, h_u), yhat_univ_2)
    expect_equal(GaussianNadarayaWatsonEstimator(x, y, h_m, xargs), yhat_multi_1)
    expect_equal(GaussianNadarayaWatsonEstimator(x, y, h_m), yhat_multi_2)
})


tests/testthat/test-TrimmingIndicator.R
test_that("When a matrix is given, it should return the product of trimming indicators of its columns.", {

    # data generating process
    N <- 500L
    x1 <- stats::rnorm(N)
    x2 <- stats::rt(N, df = 2)^2
    x3 <- stats::runif(N)
    x <- cbind(x1, x2, x3)

    # trimming indicator from scratch for a matrix
    alpha <- 0.025
    trimming_indicator <- rep(1L, N)
    for (k in 1:ncol(x)){

        w <- x[,k]
        cutoffs <- stats::quantile(w, c(alpha, 1-alpha))
        trimming_indicator <- trimming_indicator * ifelse(w > cutoffs[1L] & w < cutoffs[2L], 1L, 0L)
    }

    expect_equal(TrimmingIndicator(x, alpha), trimming_indicator)
})


MonteCarlo.R
# library(semiBRM)

hhmmss <- function(seconds){
    hh = seconds%/%3600
    mm = (seconds - hh*3600)%/%60
    ss = seconds - hh*3600 - mm*60
    return(sprintf("%02.0f:%02.0f:%02.0f", hh, mm, ss))
}


# marginal effects ----------------------------------------------------------------------------

S <- 300L

N <- 1000L
beta <- c(2, 2, -1, -1)

coef <- matrix(NaN, nrow = S, 2L)
coef.stder <- matrix(NaN, nrow = S, 2L)

avme.true <- matrix(NaN, nrow = S, 1L)
avme.semi <- matrix(NaN, nrow = S, 1L)
avme.stder <- matrix(NaN, nrow = S, 1L)
avme.error <- matrix(NaN, nrow = S, 1L)

qme.true <- matrix(NaN, nrow = S, 3L)
qme.semi <- matrix(NaN, nrow = S, 3L)
qme.stder <- matrix(NaN, nrow = S, 3L)
qme.error <- matrix(NaN, nrow = S, 3L)

start <- proc.time()

for (s in seq_len(S)){

    X1 <- rnorm(N)
    X2 <- (X1 + 2*rnorm(N))/sqrt(5) + 1
    X3 <- rnorm(N)^2/sqrt(2)
    X <- cbind(X1, X2, X3)
    V <- as.vector(cbind(X, 1)%*%beta)
    Y <- ifelse(V >= rnorm(N), 1L, 0L)

    qmle <- semiBRM(x = X, y = Y, control = list(iterlim=50))

    # coefficients
    coef[s,] <- coef(qmle)
    coef.stder[s,] <- sqrt(diag(vcov(qmle)))


    # average marginal effects
    me <- pnorm(as.vector(cbind(X1+sd(X1), X2, X3, 1)%*%beta)) -
        pnorm(as.vector(cbind(X, 1)%*%beta))

    av_me_true <- mean(me)

    av_me_semi <-  MarginalEffects(qmle, variable = "X1", delta = sd(X1))

    avme.true[s,] <- av_me_true
    avme.semi[s,] <- av_me_semi[1L]
    avme.stder[s,] <- av_me_semi[2L]
    avme.error[s,] <- av_me_semi[1L] - av_me_true

    # quantile marginal effects
    x1.cutoff <- quantile(X1, c(1/3, 2/3))

    g1_true <- mean(me[ X1 <= x1.cutoff[1L]])
    g2_true <- mean(me[ X1 >  x1.cutoff[1L] & X1 <= x1.cutoff[2L] ])
    g3_true <- mean(me[ X1 >  x1.cutoff[2L] ])

    q_me_semi <-  MarginalEffects(qmle, variable = "X1", delta = sd(X1), p.cutoffs = c(1/3, 2/3))

    qme.true[s,] <-  c(g1_true, g2_true, g3_true)
    qme.semi[s,] <- q_me_semi[,1L]
    qme.stder[s,] <- q_me_semi[,2L]
    qme.error[s,] <- q_me_semi[,1L] - c(g1_true, g2_true, g3_true)

    cat(sprintf("progress = %03d/%03d done (elapsed = %s)\r", s, S, hhmmss( (proc.time()-start)[3L] )))

    if (s==S) cat("\n");
}


# without trimming ----------------------------------------------------------------------------

# matrixStats::colMeans2(avme.true)
# [1] 0.2005231
# matrixStats::colMeans2(avme.semi)
# [1] 0.1940123
# matrixStats::colMeans2(avme.stder)
# [1] 0.006110306
# matrixStats::colSds(avme.error)
# [1] 0.00696816
# matrixStats::colMeans2(avme.error)
# [1] -0.006510768

# matrixStats::colMeans2(qme.true)
# [1] 0.24908598 0.27697798 0.07535945
# matrixStats::colMeans2(qme.semi)
# [1] 0.23737675 0.26203671 0.08249326
# matrixStats::colMeans2(qme.stder)
# [1] 0.010624265 0.010250780 0.007762536
# matrixStats::colSds(qme.error)
# [1] 0.012519107 0.011962739 0.004470873
# matrixStats::colMeans2(qme.error)
# [1] -0.011709231 -0.014941269  0.007133807



# point estimation: semiparametric bootstrapping ----------------------------------------------

S <- 300L

N <- 1200L
beta <- c(2, 2, -1, -1)
Xbar0 <- Xbar1 <- t(c(0, 1, 1/sqrt(2)))
Xbar1[,1L] <- Xbar1[,1L] + 1

prob0 <- pnorm(as.vector(cbind(Xbar0, 1)%*%beta))
prob1 <- pnorm(as.vector(cbind(Xbar1, 1)%*%beta))
me.true <- prob1 - prob0

probs_coef_semi <- rep(NaN, S)
probs_coef_boot <- rep(NaN, S)
probs_stde_boot <- rep(NaN, S)
probs_conf_boot <- matrix(NaN, S, 2)

probs_stde_pack <- rep(NaN, S)
probs_conf_pack <- matrix(NaN, S, 2)

me_coef_semi <- rep(NaN, S)
me_coef_boot <- rep(NaN, S)
me_stde_boot <- rep(NaN)

start <- proc.time()
for (s in seq_len(S)){

    X1 <- rnorm(N)
    X2 <- (X1 + 2*rnorm(N))/sqrt(5) + 1
    X3 <- rnorm(N)^2/sqrt(2)
    X <- cbind(X1, X2, X3)
    V <- as.vector(cbind(X, 1)%*%beta)
    Y <- ifelse(V >= rnorm(N), 1L, 0L)

    # parameter fit
    qmle <- semiBRM(x = X, y = Y, control = list(iterlim=50))

    # point estimation of conditional probability
    prob.semi <- predict(qmle, newdata = Xbar0, boot.se = TRUE)

    # point estimation of marginal effect as difference
    Vhat <- X[,1L] + as.vector(X[,-1L]%*%coef(qmle))
    h <- sd(Vhat)*1.06*N^(-1/5)

    V0 <- Xbar0[,1L] + as.vector(Xbar0[,-1L]%*%coef(qmle))
    V1 <- Xbar1[,1L] + as.vector(Xbar1[,-1L]%*%coef(qmle))

    me.semi <- GaussianNadarayaWatsonEstimator(Vhat, Y, h, V1) -
        GaussianNadarayaWatsonEstimator(Vhat, Y, h, V0)

    ## semiparametric bootstrapping
    B <- 500L
    prob_boot <- rep(NaN, B)
    me_boot <- rep(NaN, B)
    for (b in seq_len(B)){

        prob0.semi <- predict(qmle, newdata = Xbar0)$prob
        prob1.semi <- predict(qmle, newdata = Xbar1)$prob

        Y0.boot <- ifelse(runif(N) <= prob0.semi, 1L, 0L)
        Y1.boot <- ifelse(runif(N) <= prob1.semi, 1L, 0L)

        prob0.boot <- GaussianNadarayaWatsonEstimator(Vhat, Y0.boot, h, V0)
        prob1.boot <- GaussianNadarayaWatsonEstimator(Vhat, Y1.boot, h, V1)

        prob_boot[b] <- prob0.boot
        me_boot[b] <- prob1.boot - prob0.boot
    }

    probs_coef_semi[s] <- prob.semi$prob
    probs_coef_boot[s] <- mean(prob_boot)
    probs_stde_boot[s] <- sd(prob_boot)
    probs_conf_boot[s,] <- quantile(prob_boot, c(0.025, 0.975))

    probs_stde_pack[s] <- prob.semi$boot.se
    probs_conf_pack[s,] <- prob.semi$boot.ci[1L,]

    me_coef_semi[s] <- me.semi
    me_coef_boot[s] <- mean(me_boot)
    me_stde_boot[s] <- sd(me_boot)


    cat(sprintf("progress = %03d/%03d done (elapsed = %s)\r", s, S, hhmmss( (proc.time()-start)[3L] )))
    if (s == S) cat("\n");
}

prob0
mean(probs_coef_semi-prob0)
mean(probs_coef_semi)
mean(probs_coef_boot)

sd(probs_coef_semi)
mean(probs_stde_boot)
mean(probs_stde_pack)

colMeans(probs_conf_boot)
colMeans(probs_conf_pack)

me.true
mean(me_coef_semi-me.true)
mean(me_coef_semi)
mean(me_coef_boot)

sd(me_coef_semi)
mean(me_stde_boot)



# parameter fit
qmle <- semiBRM(x = X, y = Y, control = list(iterlim=50))

# point estimation of conditional probability
prob.semi_test <- predict(qmle, newdata = X[1:5,], boot.se = TRUE)

prob.semi_insm <- predict(qmle, boot.se = TRUE)



# point estimation: parameter simulation ------------------------------------------------------

S <- 300L

N <- 1200L
beta <- c(2, 2, -1, -1)

Xbar <- t(c(0, 1, 1/sqrt(2)))

probs_semi <- rep(NaN, S)
error_semi <- rep(NaN, S)
stder_semi <- rep(NaN, S)


start <- proc.time()
for (s in seq_len(S)){

    X1 <- rnorm(N)
    X2 <- (X1 + 2*rnorm(N))/sqrt(5) + 1
    X3 <- rnorm(N)^2/sqrt(2)
    X <- cbind(X1, X2, X3)
    V <- as.vector(cbind(X, 1)%*%beta)
    Y <- ifelse(V >= rnorm(N), 1L, 0L)

    qmle <- semiBRM(x = X, y = Y, control = list(iterlim=50))

    prob.true <- pnorm(as.vector(cbind(Xbar, 1)%*%beta))
    prob.semi <- predict(qmle, newdata = Xbar)

    Vhat <- X[,1L] + as.vector(X[,-1L]%*%coef(qmle))
    h <- sd(Vhat)*N^(-1/5)

    parms <- cbind(1, mvrnorm(n = 100, mu=coef(qmle), Sigma=vcov(qmle)))
    Vnew <- as.vector(tcrossprod(parms, Xbar))

    prob.simul <- GaussianNadarayaWatsonEstimator(Vhat, Y, h, Vnew)

    probs_semi[s] <- prob.semi$prob
    stder_semi[s] <- sd(prob.simul)
    error_semi[s] <- prob.semi$prob - prob.true


    cat(sprintf("progress = %03d/%03d done (elapsed = %s)\r", s, S, hhmmss( (proc.time()-start)[3L] )))
    if (s==S) cat("\n");
}

mean(probs_semi)
mean(stder_semi)
sd(probs_semi)



R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

.GaussianMultivar <- function(Y, X, args, H) {
    .Call('_semiBRM_GaussianMultivar', PACKAGE = 'semiBRM', Y, X, args, H)
}

.GaussianMultivarOMP <- function(Y, X, args, H, n_cores = 3L) {
    .Call('_semiBRM_GaussianMultivarOMP', PACKAGE = 'semiBRM', Y, X, args, H, n_cores)
}

.GaussianMultivarLeaveOneOut <- function(Y, X, H) {
    .Call('_semiBRM_GaussianMultivarLeaveOneOut', PACKAGE = 'semiBRM', Y, X, H)
}

.GaussianMultivarLeaveOneOutOMP <- function(Y, X, H, n_cores = 3L) {
    .Call('_semiBRM_GaussianMultivarLeaveOneOutOMP', PACKAGE = 'semiBRM', Y, X, H, n_cores)
}

.GaussianUnivar <- function(Y, X, args, H) {
    .Call('_semiBRM_GaussianUnivar', PACKAGE = 'semiBRM', Y, X, args, H)
}

.GaussianUnivarOMP <- function(Y, X, args, H, n_cores = 3L) {
    .Call('_semiBRM_GaussianUnivarOMP', PACKAGE = 'semiBRM', Y, X, args, H, n_cores)
}

.GaussianUnivarLeaveOneOut <- function(Y, X, H) {
    .Call('_semiBRM_GaussianUnivarLeaveOneOut', PACKAGE = 'semiBRM', Y, X, H)
}

.GaussianUnivarLeaveOneOutOMP <- function(Y, X, H, n_cores = 3L) {
    .Call('_semiBRM_GaussianUnivarLeaveOneOutOMP', PACKAGE = 'semiBRM', Y, X, H, n_cores)
}



Example.R
# setup ---------------------------------------------------------------------------------------
library(semiBRM)
set.seed(20190815) # for reproduction of results

# data generating process ----------------------------------------------------------------------

N <- 1500L
X1 <- rnorm(N)
X2 <- (X1 + 2*rnorm(N))/sqrt(5) + 1
X3 <- rnorm(N)^2/sqrt(2)
X <- cbind(X1, X2, X3)
beta <- c(2, 2, -1, -1)
V <- as.vector(cbind(X, 1)%*%beta)
Y <- ifelse(V >= rnorm(N), 1L, 0L)


# parameter estimation -------------------------------------------------------------------------

## estimands: the rescaled coefficients by the first coefficient excluding intercept
coefs_true <- c(1, -.5)
data <- data.frame(Y, X1, X2, X3)

## Klein and Spady (1993): semiparametric approach
semi <- semiBRM(Y ~ X1 + X2 + X3, data = data, control = list(iterlim=50))
coefs_semi <- coef(semi)

## Probit: parametric approach
probit <- glm(Y ~ X1 + X2 + X3, family = binomial(link = "probit"), data = data)
coefs_probit <- probit$coefficients[-1L][-1L]/probit$coefficients[-1L][1L]

## formatted print
{
    cat(sprintf("        %7s %7s %7s\n", "True", "Probit", "Semi"))
    cat(sprintf("parm 1: %7.4f %7.4f %7.4f\n", coefs_true[1L], coefs_probit[1L], coefs_semi[1L]))
    cat(sprintf("parm 2: %7.4f %7.4f %7.4f\n", coefs_true[2L], coefs_probit[2L], coefs_semi[2L]))
}


# in-sample prediction ------------------------------------------------------------------------

## in-sample conditional probability
in_prob_true <- pnorm(V)
in_prob_semi <- predict(semi)
in_prob_probit <- fitted(probit)

## formatted print
target <- sample.int(N, size = 10L)
{
    cat(sprintf("%7s %8s %8s %8s %12s\n", "Obs.", "True", "Probit", "Semi", "non.endpoint") )
    for (i in target){
        cat(sprintf("[%04d]: %8.6f %8.6f %8.6f %12s\n",
                    i, in_prob_true[i], in_prob_probit[i], in_prob_semi$prob[i], in_prob_semi$non.endpoint[i]))
    }
}


# out-of-sample prediction --------------------------------------------------------------------

## conditional probability at the means
Xbar <- colMeans(X)
newdata <- as.data.frame(as.list(Xbar))

## predictions
out_prob_true <- pnorm(as.vector(c(Xbar, 1)%*%beta))
out_prob_semi <- predict(semi, newdata, boot.se = TRUE)
out_prob_probit <- pnorm(as.vector(coef(probit)%*%c(1, Xbar)))

## standard errors of Probit
grad <- dnorm(as.vector(coef(probit)%*%c(1, Xbar))) * c(1, Xbar)
out_stde_probit <- sqrt(as.vector(crossprod(grad, vcov(probit))%*%grad))

## formatted print
{
    cat(sprintf("            %7s %7s %7s\n", "True", "Probit", "Semi"))
    cat(sprintf("Prob. Est.: %7.4f %7.4f %7.4f\n", out_prob_true, out_prob_probit, out_prob_semi$prob))
    cat(sprintf(" Std. Err.: %7s %7.4f %7.4f\n", "", out_stde_probit, out_prob_semi$boot.se))
}

# average marginal effects --------------------------------------------------------------------

## marginal effects as difference betwen conditional probabilities with and without perturbation
delta <- sd(X1) # size of perturbation

me_true <- pnorm(as.vector(cbind(X1+delta, X2, X3, 1)%*%beta)) - pnorm(as.vector(cbind(X, 1)%*%beta))

## average marginal effects
av_me_true <- mean(me_true)
av_me_semi <- MarginalEffects(semi, variable = "X1", delta = delta)

## formatted print
{
    cat(sprintf("           %7s %7s \n", "True", "Semi"))
    cat(sprintf("Marg.Eff.: %7.4f %7.4f\n", av_me_true, av_me_semi[1L]))
    cat(sprintf("Std. Err.: %7s %7.4f\n", "", av_me_semi[2L]))
}


# quantitle marginal effects ------------------------------------------------------------------

## percentile cutoffs
p.cutoffs <- c(1/4, 2/4, 3/4)

## group indicators
q_vals <- quantile(X1, p.cutoffs)
q1 <- ifelse(X1 <= q_vals[1L], 1L, 0L)
q2 <- ifelse(q_vals[1L] < X1 & X1 <= q_vals[2L], 1L, 0L)
q3 <- ifelse(q_vals[2L] < X1 & X1 <= q_vals[3L], 1L, 0L)
q4 <- ifelse(q_vals[3L] < X1, 1L, 0L)

## quantile marginal effects
q_me_true <- c("G1" = mean(me_true[q1==1]), "G2" = mean(me_true[q2==1]),
               "G3" = mean(me_true[q3==1]), "G4" = mean(me_true[q4==1]))

q_me_semi <- MarginalEffects(semi, variable = "X1", delta, p.cutoffs)

## formatted print
{
    cat("         ", sprintf(" %5s  ", paste0("G", 1:4)), fill = TRUE)
    cat("True    :", sprintf(" %.4f ", q_me_true), fill = TRUE)
    cat("Semi    :", sprintf(" %.4f ", q_me_semi[,1L]), fill = TRUE)
    cat("Std.Err.:", sprintf("(%.4f)", q_me_semi[,2L]), fill = TRUE)
}

